### PR TITLE
Feature/spark ui visualisation

### PR DESCRIPTION
**Pull request by dJalon**
# Description

New feature added. In the Spark ui, Waimak now adds a tab displaying a DAG.

waimakdag

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

This has been tested visually to spot inaccurate DAGs displayed in the web interface. 
Several unit tests have been left written to cover the correct functionality of the new functions added.

### CODE REVIEW

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/ui/WaimakGraph.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
+package com.coxautodata.waimak.dataflow.spark.ui
+
+import com.coxautodata.waimak.dataflow.{DataFlow, DataFlowAction}
+
+
+case class WaimakEdge(fromID: Int, toID: Int) {
+  def makeDotEdge: String = s"""  $fromID->$toID;\n"""
+}
+
+//case class WaimakNode(id: Long, actionID: String, actionDesc: String, innerNodes: Seq[WaimakNode] = Seq.empty) {
```
Comment (vavison) [inline waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/ui/WaimakGraph.scala]: Remove commented-out code

**Comment by alexjbush** [comment]
Would be nice to put a screenshot and a few lines in the readme since it's such a neat feature

**Comment by vavison** [comment]
There seems to be an issue with none of the flow-related jobs appearing in the job or stage tabs of the history server (everything is fine in the Spark UI while the job is running).

Weird one - needs investigating

File: waimak-core/src/main/scala/org/apache/spark/ui/WaimakExecutionEvent.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package org.apache.spark.ui
+
+import com.coxautodata.waimak.dataflow.spark.ui.{WaimakEventListener, WaimakGraph}
+import org.apache.spark.SparkContext
+import org.apache.spark.scheduler.SparkListenerEvent
+
+case class WaimakExecutionEvent(executionId: String, actionDescriptions: Seq[String], flowGraph: WaimakGraph) extends SparkListenerEvent
```
Comment (vavison) [inline waimak-core/src/main/scala/org/apache/spark/ui/WaimakExecutionEvent.scala]: I don't think we need the actionDescriptions in here anymore as that was to build the table (which we don't need)

**Comment by alexjbush** [comment]
@alexeipab What do you think?

**Comment by alexjbush** [comment]
Closing for now, needs more thinking about how best to lay out the DAG.

**Pull request closed without merging**